metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/RestoreColumnLineageIndices.java

```diff
@@ -0,0 +1,182 @@
+import com.linkedin.metadata.entity.EntityService;
+import com.linkedin.metadata.models.AspectSpec;
+import com.linkedin.metadata.models.registry.EntityRegistry;
+import com.linkedin.metadata.search.EntitySearchService;
+import com.linkedin.metadata.search.SearchEntity;
+import com.linkedin.metadata.search.SearchResult;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.linkedin.metadata.Constants.UPSTREAM_LINEAGE_ASPECT_NAME;
+
+@Slf4j
+public class RestoreColumnLineageIndices extends UpgradeStep {
+  private static final String VERSION = "1";
+  private static final String UPGRADE_ID = "restore-column-lineage-indices";
+  private static final Integer BATCH_SIZE = 1000;
+
+  private final EntitySearchService _entitySearchService;
+  private final EntityRegistry _entityRegistry;
+
+  public RestoreColumnLineageIndices(EntityService entityService, EntitySearchService entitySearchService,
+                                EntityRegistry entityRegistry) {
+    super(entityService, VERSION, UPGRADE_ID);
+    _entitySearchService = entitySearchService;
+    _entityRegistry = entityRegistry;
+  }
+
+  @Override
+  public void upgrade() throws Exception {
+    final AuditStamp auditStamp =
+        new AuditStamp().setActor(Urn.createFromString(Constants.SYSTEM_ACTOR)).setTime(System.currentTimeMillis());
+
+    final int totalUpstreamLineageCount = getAndRestoreUpstreamLineageIndices(0, auditStamp);
+    int upstreamLineageCount = BATCH_SIZE;
+    while (upstreamLineageCount < totalUpstreamLineageCount) {
+      getAndRestoreUpstreamLineageIndices(upstreamLineageCount, auditStamp);
+      upstreamLineageCount += BATCH_SIZE;
+    }
+
+    final int totalChartInputFieldsCount = getAndRestoreInputFieldsIndices(Constants.CHART_ENTITY_NAME, 0, auditStamp);
+    int chartInputFieldsCount = BATCH_SIZE;
+    while (chartInputFieldsCount < totalChartInputFieldsCount) {
+      getAndRestoreInputFieldsIndices(Constants.CHART_ENTITY_NAME, chartInputFieldsCount, auditStamp);
+      chartInputFieldsCount += BATCH_SIZE;
+    }
+
+    final int totalDashboardInputFieldsCount = getAndRestoreInputFieldsIndices(Constants.DASHBOARD_ENTITY_NAME, 0, auditStamp);
+    int dashboardInputFieldsCount = BATCH_SIZE;
+    while (dashboardInputFieldsCount < totalDashboardInputFieldsCount) {
+      getAndRestoreInputFieldsIndices(Constants.DASHBOARD_ENTITY_NAME, dashboardInputFieldsCount, auditStamp);
+      dashboardInputFieldsCount += BATCH_SIZE;
+    }
+  }
+
+  @Nonnull
+  @Override
+  public ExecutionMode getExecutionMode() {
+    return ExecutionMode.ASYNC;
+  }
+
+  private int getAndRestoreUpstreamLineageIndices(int start, AuditStamp auditStamp)
+      throws Exception {
+    final AspectSpec upstreamLineageAspectSpec = _entityRegistry.getEntitySpec(Constants.DATASET_ENTITY_NAME)
```

Comment:
Do we need this AspectSpec? Is there any other method on entityService to produce MCL that doesn't require the AspectSpec?

Reply:
hmm all the methods I'm seeing from entityService that produce an MCL use an aspect spec. is there something specific you had in mind/know about?